Implement composer

wenzhengjiang opened the pull request:
#1128 
This is quite large PR containing below major changes:
1. Appending composer system prompt for copilot-plus mode.
2. Rendering composer markdown blocks by adding a note path title and the apply button.
3. Implement apply function by calling the composer apply endpoint and creating the Apply view.

logancyang commented on src/plusUtils.ts:
```diff
@@ -80,3 +93,28 @@ export function turnOffPlus(): void {
+  try {
+    // Get the composer prompt from the API
+    const composerPromptResponse = await brevilabsClient.composerPrompt();
+    cachedComposerPrompt = composerPromptResponse.prompt;
```
Should we persist this on disk just like how `/pdf4llm` cache works? Just need the simplest invalidation mechanism for now. Ideally we need every way possible to reduce the lag for the "Apply" action.

wenzhengjiang replied:
I won't worry about the extra latency of call. /broca plus /chat/completion take ~5000ms, while this prompt fetch will be no more than 50ms.

zeroliu replied:
Is there any reason why this prompt needs to come from the server? I assume the server doesn't do any extra processing as it doesn't take any argument from the client?

wenzhengjiang replied:
The main benefit is faster deployment. At least in the pre-release stage, I expect there will be a lot of prompt-tuning and deploying changes from the backend is a lot easier than asking users to upgrade their code.

After the official release, I don't have strong opinion whether putting the prompt in the backend or client side.

logancyang reviewed (commented):
Awesome job 🔥 Congrats! 🎉

Some quick observations:

- The full path and the buttons in this top section looks pretty big overall, might be better to make them smaller?
![SCR-20250303-ppzt](https://github.com/user-attachments/assets/95f76411-4745-4c25-af31-7f7a95f33bba)

- After clicking "Apply", there's a noticeable lag before the diff view is opened. It's fine to have a lag but it would be a nice UX improvement to **add a spinner** on the Apply button so that it doesn't look like nothing is happening for a second after clicking it.

- When there's nothing accepted, I click "Apply Accepted Changes" it just closes the diff view without anything changed. This is reasonable function-wise but a bit awkward on the UX side - user might think it's "Apply All" or they didn't realize they didn't accept anything. Is it better to have an "Apply all / or accepted changes", if nothing is accepted explicitly, it just acts like "Apply all"; if anything accepted, it applies those. Open to other ideas.

- When there's only one diff block with Accept / Reject, is it better to make the block-level Accept / Reject act like the global ones and close the diff view? It seems unnecessary to click the top ones again for a single block.

WDYT? @wenzhengjiang

zeroliu commented:
> * user might think it's "Apply All" or they didn't realize they didn't accept anything. Is it better to have an "Apply all / or accepted changes", if nothing is accepted explicitly, it just acts like "Apply all"; if anything accepted, it applies those. Open to other ideas.

I was thinking the exact same thing when I first saw the demo. The UX of accepting diff has been pretty standardized. The same pattern has been used not only for the latest AI editing, but also has been like that since the git diff experience. I would highly recommend us to mimic a similar experience.

zeroliu commented on package.json:
```diff
@@ -98,12 +98,14 @@
     "@radix-ui/react-switch": "^1.1.1",
     "@radix-ui/react-tooltip": "^1.1.6",
     "@tabler/icons-react": "^2.14.0",
+    "@types/diff": "^7.0.1",
```
this should be a dev dependency

wenzhengjiang replied:
Done.

zeroliu commented on src/LLMProviders/brevilabsClient.ts:
```diff
@@ -194,4 +215,12 @@ export class BrevilabsClient {
     return this.makeRequest<Youtube4llmResponse>("/youtube4llm", { url });
   }
+
+  async composerPrompt(): Promise<ComposerPromptResponse> {
+    return this.makeRequest<ComposerPromptResponse>("/composer/prompt", {}, "GET");
+  }
+
+  async composerApply(request: ComposerApplyRequest): Promise<ComposerApplyResponse> {
+    return this.makeRequest<ComposerApplyResponse>("/composer/apply", request);
+  }
```
The latest brevilabs client changed makeRequest to return both data and error. The wrapper functions here need to handle the error too

wenzhengjiang replied:
Done.

zeroliu commented on src/styles/tailwind.css:
```diff
@@ -872,3 +896,125 @@ If your plugin does not need CSS, delete this file.
 }
+
+/* Apply View Styles */
+.apply-view-container {
```
We already standardized styling to tailwindcss. All existing CSS class names only exist for legacy reasons. No changes should be introduced to tailwind.css in this PR.

wenzhengjiang replied:
Got it. Sounds good! Removed all custom CSS class below.

zeroliu commented on src/components/chat-components/ChatSingleMessage.tsx:
```diff
@@ -203,6 +296,85 @@ const ChatSingleMessage: React.FC<ChatSingleMessageProps> = ({
+        if (!contentRef.current) return;
+
+        // Find all pre elements (code blocks) in the rendered content
+        const codeBlocks = contentRef.current.querySelectorAll("pre");
```
interacting with the raw DOM element in the react component can be very error-prone. Is there any reason why it needed to be implemented this instead of relying on React functions?

wenzhengjiang replied:
No specific reason, simply never did web frontend before :). Updated the code by creating a custom React component. Hope that makes more sense to you.

wenzhengjiang commented:
Thanks for the good UX feedbacks! 
* For the top section, I also plan to remove the status bar. I feel like it's not adding much value.
* For the top apply button,  let me rename the "Apply Accepted Changes" button to "Apply changes"? And it will apply accepted or unselected changes.

> When there's only one diff block with Accept / Reject, is it better to make the block-level Accept / Reject act like the global ones and close the diff view? It seems unnecessary to click the top ones again for a single block.
Or they could just click the top ones since there is only one block? Not sure we want to change the behavior of the buttons for this specific case.

zeroliu commented on src/components/ApplyView.tsx:
```diff
@@ -0,0 +1,351 @@
+            <div
+              key={blockIndex}
+              className={cn("diff-block mb-4 border rounded-md overflow-hidden", {
+                "border-green-500 accepted": isAccepted,
+                "border-red-500 rejected": isRejected,
+                "border-gray-300": !isAccepted && !isRejected,
```
tokens like green-500, red-500, gray-300 are not valid tokens in our design system. We should use the ones configured in the design system.

wenzhengjiang replied:
I see. Done.

zeroliu commented on src/styles/tailwind.css:
```diff
@@ -872,3 +896,125 @@ If your plugin does not need CSS, delete this file.
+  margin-bottom: 12px;
+}
+
+.apply-view-summary .text-green-600 {
```
We do not allow adding classnames like this. This breaks how we work with tailwindcss. There are existing tokens that work for the color that you need. I highly recommend take a close look at tailwind.config.js and follow tailwindcss styles.

wenzhengjiang replied:
Got it. Migrates the tokens to tailwindcss.

wenzhengjiang commented:
@logancyang @zeroliu I think I have addressed all your concerns. I'm pretty new to web frontend, so I could be still missing something...

Currently I'm using the default variant for accept buttons and secondary variant for reject buttons. They look like below:

<img width="905" alt="Screenshot 2025-03-04 at 22 39 22" src="https://github.com/user-attachments/assets/310eada2-8ae8-4db6-a350-16aecf377a2b" />

Open to making the green and red changes if you think they’re worth it.

zeroliu commented on src/components/ApplyView.tsx:
```diff
@@ -0,0 +1,297 @@
+    return "Apply Changes";
+  }
+
+  async setState(state: ApplyViewState) {
+    this.state = state;
+    this.render();
+  }
```
is this function used somewhere?

zeroliu commented on src/components/ApplyView.tsx:
```diff
@@ -0,0 +1,297 @@
+
+    const container = this.containerEl.children[1];
+    container.empty();
+    container.addClass("apply-view-container");
```
do we need this class?

wenzhengjiang replied:
Ah, you are right. We don't need this anymore.

zeroliu commented on src/components/ApplyView.tsx:
```diff
@@ -0,0 +1,297 @@
+  private render() {
+    if (!this.state) return;
+
+    const container = this.containerEl.children[1];
```
why `this.containerEl.children[1];`? Is this deterministic? what other children does the container element have?

wenzhengjiang replied:
Didn't find any official documentation, but this seems like a standard pattern to access the second child and the content view while the first child as the header view. See https://github.com/search?q=this.containerEl.children%5B1%5D&type=code

zeroliu commented on src/components/ApplyView.tsx:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    this.root.render(
+      <ApplyViewRoot app={this.app} state={this.state} close={() => this.leaf.detach()} />
```
I don't see how this.state is set. `setState` is not called in this PR and it's not coming from constructor.

wenzhengjiang replied:
It's set in https://github.com/logancyang/obsidian-copilot/pull/1315/files#diff-eee12cc355c7066cd49d0c5b30e8b3ec2ad7c87acc76702adae3da7a9cd6b54dR150 via setViewState of the WorkspaceLeaf object. I think internally WorkspaceLeaf called the setState to pass the state object. 

Again, not official document, but found something related in https://liamca.in/Obsidian/API+FAQ/views/persisting+your+view+state.

zeroliu commented on src/components/ApplyView.tsx:
```diff
@@ -0,0 +1,297 @@
+    setChangeBlocks(blocks);
+  }, [diff]);
+
+  // Add defensive check for state after hooks
```
How would the UI enter this state?

zeroliu commented on src/components/ApplyView.tsx:
```diff
@@ -0,0 +1,297 @@
+        if (index !== -1) {
+          newDiff[index] = {
+            ...newDiff[index],
+            accepted: true,
+            rejected: false,
```
why do we need two flags to represent this value? is there a case where accepted and rejected can be same values?

zeroliu commented on src/components/ApplyView.tsx:
```diff
@@ -0,0 +1,297 @@
+                }
+              )}
+            >
+              {block.map((change, changeIndex) => (
+                <div
+                  key={`${blockIndex}-${changeIndex}`}
+                  className={cn("flex relative border-l-3", {
+                    "bg-modifier-success border-l-green": change.added,
+                    "bg-modifier-error border-l-red": change.removed,
+                    "opacity-50":
+                      (change.added && change.rejected) || (change.removed && !change.accepted),
+                  })}
+                >
+                  <div className="w-6 flex-shrink-0 flex items-center justify-center text-sm font-bold bg-background-secondary-alt">
+                    {change.added && "+"}
+                    {change.removed && "-"}
+                  </div>
+                  <div className="flex-1 font-mono text-sm whitespace-pre-wrap py-1 px-2 text-text-normal">
+                    {change.value}
+                  </div>
+                </div>
+              ))}
+
+              {/* Only show accept/reject buttons for blocks with changes */}
+              {hasChanges && (
+                <div className="flex justify-end p-2 bg-background-secondary-alt border-t border-border gap-2">
+                  <Button variant="secondary" size="sm" onClick={() => rejectBlock(blockIndex)}>
+                    <XIcon className="h-4 w-4" />
+                    Reject
+                  </Button>
+                  <Button size="sm" onClick={() => acceptBlock(blockIndex)}>
+                    <Check className="h-4 w-4" />
+                    Accept
+                  </Button>
+                </div>
+              )}
+            </div>
```
can we apply the change directly in this view when a user accept or reject a block? It's not very clear after a user accepts and rejects the change and what's going to happen

zeroliu commented on src/plusUtils.ts:
```diff
@@ -80,3 +93,28 @@ export function turnOffPlus(): void {
   }
 }
+
+export async function getComposerSystemPrompt(): Promise<string> {
```
I suggest move anything composer related to its own file. `plusUtil` should only take care of plus user status checks

zeroliu commented on src/components/chat-components/ChatSingleMessage.tsx:
```diff
@@ -79,6 +85,100 @@ const ChatSingleMessage: React.FC<ChatSingleMessageProps> = ({
     });
   };
 
+  const handleApplyCode = useCallback(
```
I would also suggest create a custom hook and move this chunk of code to a separate module under composer. Making composer code more isolated from the core chat.

zeroliu commented on src/components/chat-components/ChatSingleMessage.tsx:
```diff
@@ -79,6 +85,100 @@ const ChatSingleMessage: React.FC<ChatSingleMessageProps> = ({
+            type: "obsidian-copilot-apply-view",
+            active: true,
+            state: {
+              file: file,
+              originalContent: originalContent,
+              newContent: newContent,
+              path: path,
```
```suggestion
              file,
              originalContent,
              newContent,
              path,
```

zeroliu commented on src/components/chat-components/ChatSingleMessage.tsx:
```diff
@@ -203,6 +303,32 @@ const ChatSingleMessage: React.FC<ChatSingleMessageProps> = ({
       );
+
+      // Process code blocks after rendering
+      const codeBlocks = contentRef.current.querySelectorAll("pre");
```
is there any reason why we need to escape from react component tree here?

zeroliu commented on src/styles/tailwind.css:
```diff
@@ -432,26 +432,50 @@ If your plugin does not need CSS, delete this file.
 .message-content pre .copy-code-button {
```
can you please help me understand why we need the changes in this file still?

zeroliu reviewed (commented):
looks much better. Left more comment as I'm digging into more implementation details

zeroliu commented:
> @logancyang @zeroliu I think I have addressed all your concerns. I'm pretty new to web frontend, so I could be still missing something...
> 
> Currently I'm using the default variant for accept/apply buttons and secondary variant for reject buttons. They look like below:
> 
> <img alt="Screenshot 2025-03-04 at 22 39 22" width="905" src="https://private-user-images.githubusercontent.com/3186972/419038018-310eada2-8ae8-4db6-a350-16aecf377a2b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEzMjAyMjksIm5iZiI6MTc0MTMxOTkyOSwicGF0aCI6Ii8zMTg2OTcyLzQxOTAzODAxOC0zMTBlYWRhMi04YWU4LTRkYjYtYTM1MC0xNmFlY2YzNzdhMmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDdUMDM1ODQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmE0MjRhNGUxYmQyNTA0OGVlN2Y2Nzk2MzQ2MWI4ZTIwYWNhZDMwZWViMDAxYzJhOWI5MjczYmM1MDA2NzMyZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.spffHQJaa_jv4mN2NC60-6GekSvv63zE1e7g2VsGs4E">
> Open to making the green and red changes if you think they’re worth it.

We will probably need new tokens for this since it's a new type of UI. I can help you create them once we get the main body of the code ready 🥂

wenzhengjiang commented:
> looks much better. Left more comment as I'm digging into more implementation details

Thanks for the detailed review! Latest updates:
1. Isolate composer code to their own files
2. Update the UI to use checkbox to select change blocks.
3. Revert all changes on classical css classes.


<img width="726" alt="Screenshot 2025-03-07 at 23 18 45" src="https://github.com/user-attachments/assets/10fd12b7-627d-4daa-a376-922bd3924cdb" />

logancyang commented on src/components/composer/ApplyView.tsx:
```diff
@@ -0,0 +1,437 @@
+
+  // Add refs to track change blocks
+  const blockRefs = useRef<(HTMLDivElement | null)[]>([]);
+  console.log(changeBlocks);
```
Should this be removed

logancyang commented:
Another observation is that I always get a single empty line diff at the end every time I hit Apply, might worth removing it programmatically? @wenzhengjiang @zeroliu 

<img width="563" alt="SCR-20250311-kdoe" src="https://github.com/user-attachments/assets/4fbb26cd-71d8-4d8c-b875-73abbf8e36f2" />

logancyang reviewed (approved):
LGTM!! This is a huge milestone, congrats everyone! 🚀🚀🚀

wenzhengjiang commented:
Plan for improving the current implementation:
1. Trigger composer with the `@composer` tool
2. The `@composer` tool:
  * Input: user query, chat history, original content of target notes
  * Output: diff in md block, full new content.
  * Can call the current chat model, but should be in a new chat session.
  * diff md only needs to contain the added lines. (for now)
3. diff md block is added to chat context to be displayed to users.
4. The full new content is added to a shared state to be fetched in apply view.

@logancyang @zeroliu @Emt-lin
